File: .a-new-structure/b-word/L0-universal/c-hybrid/config/src/formats.rs
//! Standard format detection from file extensions.
//!
//! Ported from Go `formats.go`. L0 knows standard formats (.go, .toml, .json, etc.).
//! L1 adds OmniCode-specific extensions on top (.omni, .gen, .witness, etc.)
//! via [`register_format`] / [`register_config_format`].
//!
//! Genesis 1:4 — "And God divided the light from the darkness."
//! Format detection separates — each file gets its proper identity.

use std::collections::{BTreeMap, BTreeSet};
use std::path::Path;
use std::sync::{LazyLock, RwLock};

// ────────────────────────────────────────────────────────────────
// Format Registries
// ────────────────────────────────────────────────────────────────

/// Standard file extensions → format identifiers.
///
/// Populated at init with universal formats. L1 extends via [`register_format`].
/// Uses `RwLock` for thread-safe registration (reads are concurrent, writes exclusive).
static STANDARD_FORMATS: LazyLock<RwLock<BTreeMap<String, String>>> = LazyLock::new(|| {
    let map = BTreeMap::from([
        // Compiled languages
        (".go".into(), "go".into()),
        (".c".into(), "c".into()),
        (".h".into(), "c".into()),
        (".rs".into(), "rs".into()),
        // Config/data formats
        (".toml".into(), "toml".into()),
        (".json".into(), "json".into()),
        (".jsonc".into(), "jsonc".into()),
        (".yaml".into(), "yaml".into()),
        (".yml".into(), "yaml".into()),
        // Documentation
        (".md".into(), "md".into()),
        (".adoc".into(), "adoc".into()),
        (".txt".into(), "txt".into()),
        // Scripting
        (".py".into(), "py".into()),
        (".ts".into(), "ts".into()),
        (".js".into(), "js".into()),
        (".sh".into(), "sh".into()),
        (".bash".into(), "sh".into()),
        // Web
        (".html".into(), "html".into()),
        (".css".into(), "css".into()),
    ]);
    RwLock::new(map)
});

/// Formats that hold configuration data.
///
/// Used by [`is_config_format`] and [`is_config_path`] to identify loadable configs.
static CONFIG_FORMATS: LazyLock<RwLock<BTreeSet<String>>> = LazyLock::new(|| {
    let set = BTreeSet::from(["toml".into(), "json".into(), "jsonc".into(), "yaml".into()]);
    RwLock::new(set)
});

// ────────────────────────────────────────────────────────────────
// Registration — L1 extends L0 without modifying it
// ────────────────────────────────────────────────────────────────

/// Add a file extension → format mapping.
///
/// L1 calls this to register OmniCode-specific extensions (.omni, .gen, etc.)
/// without modifying L0's source. Call during initialization.
///
/// `ext` must include the dot: `".omni"`, not `"omni"`.
pub fn register_format(ext: &str, format: &str) {
    let mut map = STANDARD_FORMATS.write().unwrap_or_else(|e| e.into_inner());
    map.insert(ext.to_owned(), format.to_owned());
}

/// Mark a format as holding configuration data.
///
/// After registration, [`is_config_format`] and [`is_config_path`] recognize it.
pub fn register_config_format(format: &str) {
    let mut set = CONFIG_FORMATS.write().unwrap_or_else(|e| e.into_inner());
    set.insert(format.to_owned());
}

// ────────────────────────────────────────────────────────────────
// Format Detection
// ────────────────────────────────────────────────────────────────

/// Returns the format identifier for a file extension, or `None` if unrecognized.
///
/// Checks both L0 standard formats and any L1-registered extensions.
pub fn format_from_ext(ext: &str) -> Option<String> {
    let map = STANDARD_FORMATS.read().unwrap_or_else(|e| e.into_inner());
    map.get(ext).cloned()
}

/// Extracts the file extension from `path` and returns its format identifier.
pub fn format_from_path(path: &Path) -> Option<String> {
    let ext = path.extension()?.to_str()?;
    let dotted = format!(".{ext}");
    format_from_ext(&dotted)
}

/// Returns true if the format holds configuration data.
pub fn is_config_format(format: &str) -> bool {
    let set = CONFIG_FORMATS.read().unwrap_or_else(|e| e.into_inner());
    set.contains(format)
}

/// Returns true if the file path has a config format extension.
pub fn is_config_path(path: &Path) -> bool {
    format_from_path(path)
        .map(|f| is_config_format(&f))
        .unwrap_or(false)
}

/// Returns true if the file extension maps to a known format.
/// Includes both L0 standard formats and L1-registered extensions.
pub fn is_known_ext(ext: &str) -> bool {
    let map = STANDARD_FORMATS.read().unwrap_or_else(|e| e.into_inner());
    map.contains_key(ext)
}

// ────────────────────────────────────────────────────────────────
// Introspection
// ────────────────────────────────────────────────────────────────

/// Returns a sorted list of all registered file extensions.
///
/// Includes both L0 standard and L1-registered extensions.
/// Useful for help text, debugging, and validation error messages.
pub fn all_known_exts() -> Vec<String> {
    let map = STANDARD_FORMATS.read().unwrap_or_else(|e| e.into_inner());
    map.keys().cloned().collect() // BTreeMap already sorted
}

/// Returns a sorted list of all registered config format names.
pub fn all_config_formats() -> Vec<String> {
    let set = CONFIG_FORMATS.read().unwrap_or_else(|e| e.into_inner());
    set.iter().cloned().collect() // BTreeSet already sorted
}

// ────────────────────────────────────────────────────────────────
// Test-only: cleanup helper
// ────────────────────────────────────────────────────────────────

/// Remove a format registration (for test isolation).
#[cfg(test)]
pub(crate) fn unregister_format(ext: &str) {
    let mut map = STANDARD_FORMATS.write().unwrap_or_else(|e| e.into_inner());
    map.remove(ext);
}

/// Remove a config format registration (for test isolation).
#[cfg(test)]
pub(crate) fn unregister_config_format(format: &str) {
    let mut set = CONFIG_FORMATS.write().unwrap_or_else(|e| e.into_inner());
    set.remove(format);
}

File: .a-new-structure/b-word/L0-universal/c-hybrid/spec/config/go/util/formats.go
//omni:code --go -library
//omni:key B-L0-config-util-formats
//omni:version b-01.00

// ============================================================================
// METADATA
// ============================================================================
//
// Key: B-L0-config-util-formats
// Purpose: Standard format detection from file extensions
// Biblical: Genesis 1:4 - "And God divided the light from the darkness"
//           Format detection separates — each file gets its proper identity.
// Authors: Nova Dawn
// Version: b-01.00
// Created: 2026-02-13
//
// L0 knows standard formats (.go, .toml, .json, etc.).
// L1 adds OmniCode-specific extensions on top (.omni, .gen, .witness, etc.).
// This separation keeps L0 universal — no OmniCode dependency.
//
// ============================================================================
// END METADATA
// ============================================================================

package util

// ============================================================================
// SETUP
// ============================================================================

import (
	"path/filepath"
	"sort"
	"sync"
)

// ────────────────────────────────────────────────────────────────
// Format Registries
// ────────────────────────────────────────────────────────────────

// mu protects StandardFormats and ConfigFormats from concurrent registration.
// Reads don't need locking — maps are populated at init and extended via
// RegisterFormat/RegisterConfigFormat before concurrent access begins.
// The lock prevents races if multiple init() functions register simultaneously.
var mu sync.Mutex

// StandardFormats maps standard file extensions to format identifiers.
// These are universal — any project would recognize them.
// OmniCode-specific extensions (.omni, .gen, etc.) belong in L1
// and are registered via RegisterFormat.
var StandardFormats = map[string]string{
	// Compiled languages
	".go": "go",
	".c":  "c",
	".h":  "c",
	".rs": "rs",

	// Config/data formats
	".toml":  "toml",
	".json":  "json",
	".jsonc": "jsonc",
	".yaml":  "yaml",
	".yml":   "yaml",

	// Documentation
	".md":   "md",
	".adoc": "adoc",
	".txt":  "txt",

	// Scripting
	".py":   "py",
	".ts":   "ts",
	".js":   "js",
	".sh":   "sh",
	".bash": "sh",

	// Web
	".html": "html",
	".css":  "css",
}

// ConfigFormats lists formats that hold configuration data.
// Used by IsConfigFormat and IsConfigPath to identify loadable configs.
var ConfigFormats = map[string]bool{
	"toml":  true,
	"json":  true,
	"jsonc": true,
	"yaml":  true,
}

// ============================================================================
// END SETUP
// ============================================================================

// ============================================================================
// BODY
// ============================================================================

// ────────────────────────────────────────────────────────────────
// Registration — L1 extends L0 without modifying it
// ────────────────────────────────────────────────────────────────

// RegisterFormat adds a file extension → format mapping.
// L1 calls this to register OmniCode-specific extensions (.omni, .gen, etc.)
// without modifying L0's source. Call during init().
//
// ext must include the dot: ".omni", not "omni".
func RegisterFormat(ext, format string) {
	mu.Lock()
	StandardFormats[ext] = format
	mu.Unlock()
}

// RegisterConfigFormat marks a format as holding configuration data.
// After registration, IsConfigFormat and IsConfigPath recognize the format.
func RegisterConfigFormat(format string) {
	mu.Lock()
	ConfigFormats[format] = true
	mu.Unlock()
}

// ────────────────────────────────────────────────────────────────
// Format Detection
// ────────────────────────────────────────────────────────────────

// FormatFromExt returns the format identifier for a file extension.
// Returns the format string if known, empty string if unrecognized.
//
// Checks both L0 standard formats and any L1-registered extensions.
func FormatFromExt(ext string) string {
	if f, ok := StandardFormats[ext]; ok {
		return f
	}
	return ""
}

// FormatFromPath extracts the file extension and returns its format identifier.
func FormatFromPath(path string) string {
	return FormatFromExt(filepath.Ext(path))
}

// IsConfigFormat returns true if the format holds configuration data.
func IsConfigFormat(format string) bool {
	return ConfigFormats[format]
}

// IsConfigPath returns true if the file path has a config format extension.
func IsConfigPath(path string) bool {
	return IsConfigFormat(FormatFromPath(path))
}

// IsKnownExt returns true if the file extension maps to a known format.
// Includes both L0 standard formats and L1-registered extensions.
func IsKnownExt(ext string) bool {
	_, ok := StandardFormats[ext]
	return ok
}

// ────────────────────────────────────────────────────────────────
// Introspection
// ────────────────────────────────────────────────────────────────

// AllKnownExts returns a sorted list of all registered file extensions.
// Includes both L0 standard and L1-registered extensions.
// Useful for help text, debugging, and validation error messages.
func AllKnownExts() []string {
	exts := make([]string, 0, len(StandardFormats))
	for ext := range StandardFormats {
		exts = append(exts, ext)
	}
	sort.Strings(exts)
	return exts
}

// AllConfigFormats returns a sorted list of all registered config format names.
func AllConfigFormats() []string {
	formats := make([]string, 0, len(ConfigFormats))
	for f := range ConfigFormats {
		formats = append(formats, f)
	}
	sort.Strings(formats)
	return formats
}

// ============================================================================
// END BODY
// ============================================================================

// ============================================================================
// CLOSING
// ============================================================================
//
// L0 format detection is deliberately limited to standard formats.
// The layering:
//
//   L0 (this file): .go → "go", .toml → "toml", .json → "json"
//   L1 (pragma):    .omni → "omni", .gen → "binary", .witness → "jsonc"
//
// L1 can call L0's FormatFromExt for standard files, and add its own
// extension registry on top. L0 never needs to know about OmniCode.
//
// "And God divided the light from the darkness." — Genesis 1:4
//
// ============================================================================
// END CLOSING
// ============================================================================
